### ISSUE TITLE
unexpected 'moodle_database' (T_STRING), expecting function (T_FUNCTION) or const (T_CONST)

### ISSUE THREAD

**Issue by ncabello0335** [donhinkelman/moodle-block_sharing_cart#111]
Hi, i have Moodle 3.9.4+ (Build: 20210226)

After i installed the plugin, when i try to add it in the course as a block i get this error:

Excepción - syntax error, unexpected 'moodle_database' (T_STRING), expecting function (T_FUNCTION) or const (T_CONST)

When i run cron i get this 
Adhoc task failed: core_course\task\course_delete_modules,Class 'block_sharing_cart\files\cleaner' not found

**Comment by donhinkelman** [comment]
Thank you for your error report. Did you install the latest build of Sharing Cart 3.10 release 6 from here on github?

**Comment by ncabello0335** [comment]
I installed the latest from moodle plugins page:  3.10, release 6 (2021072300)

**Comment by donhinkelman** [comment]
Urlik and our Praxis colleagues, this is urgent. 
<img width="501" alt="Screen Shot 2021-07-26 at 11 33 25" src="https://user-images.githubusercontent.com/12912710/126925525-323cfb4b-eba2-4a2c-80a8-87b1db3078b5.png">
I reproduced the same error after downloading the Sharing Cart 3.10, release 6 from the plugins database. This error appeared when trying to add the Sharing Cart as a block from the Add Blocks menu. I suspect this is due to one of the four pull requests that I merged three days ago. Could you look at this or assign someone? I am sorry I don't know the nature of the problem.

**Comment by amwalding** [comment]
Just FYI - got the same error here when upgrading Sharing Cart from the Moodle plugins.  Had to restore our site from a backup using the prior version of this plugin.  This update should be removed from the moodle plugins site until fixed.

**Comment by donhinkelman** [comment]
Yes, it was removed (hidden) from the Plugins Database yesterday as soon as the error was confirmed. Only 3.10, release 3 is available for download.

**Comment by Macgyverman** [comment]
The issue appears to be caused by some of the merged code requiring typed class properties, which starts in PHP 7.4 (https://www.php.net/manual/en/migration74.new-features.php). 

You can see an example in at line 30 of classes/files/cleaner.php

Previous working code:
```
class cleaner
{
    /** @var \moodle_database */
    private $db;
    /** @var file */
    private $file;
```

New merged code 
```
class cleaner
{
    private \moodle_database $db;
    private file $file;

```
Any chance you can release a new update of the plugin using the old working code? I'm one of the unfortunate users who updated to the broken code and since the last working release is an older version, I can't upgrade automatically. I also can't revert to a backup, as it's been too long since the update.

**Comment by donhinkelman** [comment]
Hi Macgyverman. Thank you very much for investigating the problem with the newly merged code. Unfortunately, I am not a programmer and I believe the Praxis team is on a well-deserved holiday now, because my messages are not returned. So perhaps you could help if you know how to make a patch and pull request. Then I will go ahead and do that. In the mean time, until this issue is resolved, I have hidden the Sharing Cart 3.10, release 6, so that only Sharing Cart 3.10, release 3 can be downloaded. I apologize to those who downloaded release 6, because you will have to either disable Sharing Cart or revert to the older version. Again, my apologies for the extra burden on you.

**Comment by Macgyverman** [comment]
Hi Don, no need to apologize... I certainly don't blame the team. I don't think I can revert versions, as such a thing doesn't appear to exist on Moodle. The only option I have is to remove the plugin, but that will also remove all data it's saved. I'm not actually familiar with what it does, so I don't want to risk deleting important data. 

I'd patch the code, but, in all honesty, I've never worked on a GitHub project, so I'm very much unsure on how you actually contribute to it. I'll have a quick read. If I can find out how to patch the code, I'll do it.

**Comment by Macgyverman** [comment]
Sorry Don, I've had a look at the Github docs and I can't honestly say I feel confident enough to know how to submit a fix to the code. I don't have enough experience with git. 

Perhaps the best thing to do would be to go back to @frederikmillingpytlick and have them fix the changes they applied to the code?

**Comment by mcardle** [comment]
Hi all...

Yes, we type all our properties in our company, because PHP 7.4 is the oldest supported version of PHP and that is what we run. I am sorry, that we released the typed properties, since there is still security fixes being made for PHP 7.3 and we should allow that being run. Also we should have mentioned it in the release notes and also made a warning like the one that is mentioned about PHP 7.2.

I will encourage all to upgrade to PHP 7.4, because in about 3 months no more security issues will fixed for PHP 7.3. See link below.

https://www.php.net/supported-versions.php
